daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/LayeredSequenceUnparser.scala

```diff
@@ -35,29 +35,55 @@ class LayeredSequenceUnparser(ctxt: SequenceRuntimeData,
   override def unparse(state: UState): Unit = {
     val layerTransformer = layerTransformerFactory.newInstance(layerRuntimeInfo)
 
-    val originalDOS = state.dataOutputStream // layer will output to the original, then finish it upon closing.
+    val originalDOS = state.dataOutputStream
+
+    // create a new buffered DOS that this layer will flush to when the layer
+    // completes unparsing. This ensures that any fragment bits or bitOrder
+    // state in the original DOS does not affect how the layer flushes bytes to
+    // the underlying DOS. Only when this new DOS is delivered to the
+    // originalDOS will the bitOrder checks be done.
+    val layerUnderlyingDOS = originalDOS.addBuffered()
+
+    // clone the UState to be used when the layer flushes its buffered content
+    // to layerUnderlyingDOS. This layer isn't a suspension, but this gives
+    // us an immutable clone that the layer can safely use
+    val finfoPre = state.asInstanceOf[UStateMain].cloneForSuspension(layerUnderlyingDOS)
+
+    // mark the original DOS as finished--no more data will be unparsed to it.
+    // If known, this will carry bit position forward to the layerUnerlyingDOS,
+    // or could even make that DOS direct. Note that we must finish with the
+    // cloned state to ensure we use this state snapshot when the
+    // layerUnderlyingDOS is delivered to it
+    originalDOS.setFinished(finfoPre)
+
+    // create a new DOS where unparsers following this layer will unparse
+    val layerFollowingDOS = layerUnderlyingDOS.addBuffered()
 
-    val newDOS = originalDOS.addBuffered() // newDOS is where unparsers after this one returns will unparse into.
-    //
     // New layerDOS is where the layer will unparse into. Ultimately anything written
-    // to layerDOS ends up, post transform, in originalDOS.
-    //
-    val layerDOS = layerTransformer.addLayer(originalDOS, state)
+    // to layerDOS ends up, post transform, in layerUnderlyingDOS
+    val layerDOS = layerTransformer.addLayer(layerUnderlyingDOS, state, finfoPre)
 
     // unparse the layer body into layerDOS
     state.dataOutputStream = layerDOS
     super.unparse(state)
     layerTransformer.endLayerForUnparse(state)
 
-    // now we're done with the layer, so finalize the layer
-    layerDOS.lastInChain.setFinished(state)
+    // now we're done with the layer, so finalize the last DOS in the chain.
+    // Note that we must reclone the UState (it probably changed since our last
+    // clone) and use that when finishing, like before, to ensures that the DOS
+    // stores a state snapshot to be used later if suspensions prevent
+    // completing right now
+    val layerDOSLast = layerDOS.lastInChain
+    val finfoPost = state.asInstanceOf[UStateMain].cloneForSuspension(layerDOSLast)
```

Comment:
When we call `setFinished(...)`, the DOS stores the state that was passed in (in the `finishedFormatInfo` variable). This state ends up being used for various things (e.g. bitOrder checks) when delivering the following DOS into it. But due to suspensions, delivering that content could be delayed and happen much later after the UState has been modified. So if we don't clone the UState, then it might change and then the state we use when we finally do deliver the content isn't correct.

Reply:
Ok. Can you add that description to the comments?

Reply:
Will do